[WIP] Review trailers tooling.

elmart opened the pull request:
Some basic tooling to support review trailers. See README for details.

@justinmk Your opinion on this is very important, as it can carry some work for you (and any other committing to master). That's because I've tried to do something so that regular users don't have to deal with trailers at all, as far as they follow a simple convention on review commits' titles. But that leaves adding Reviewed-by trailers on the shoulders of the one committing to master. I think things are left simple enough for that not to be a problem. But I'm not completely convinced. If you are not sure, proposed tools are easily modifiable so that regular users do add all trailers with little effort. But that would require regular users having to remember to do it. So, I'm open to discussion here.

elmart committed:
Review trailers tooling: Command line tool to get GitHub user info.

A tool to transform GitHub usernames into corresponding
"Full Name <email@addres>".

Put this somewhere on your $PATH.

This needs curl and jq.

elmart committed:
Review trailers tooling: Nvim plugin to insert GitHub user info.

Place this file under NVIM_RUNTIME_PATH/plugin, or copy its contents to
your .nvimrc.

After that, pressing <Leader>gu will replace word under cursor
(interpreted as a GitHub username) with corresponding info.

elmart committed:
Review trailers tooling: UltiSnips templates to insert trailer tokens.

A couple of UltiSnips templates to help inserting "Helped-by" and
"Reviewed-by" trailers.

Place this file under UltiSnips search path (usually UltiSnips dir under
Nvim's runtime path).

Snipppets activate only for git commit files, and only at the very
beginning of line, when pressing "h" or "r" (and the UltiSnips trigger).

elmart committed:
Review trailers tooling: Git hook for auto trailer insertion on rebase.

Place this at your respo's .git/hooks, and name commits introducing
changes suggested by reviewers this way:
"Review: <reviewer-github-username>: ...."

Then, you can work as always.
Whenever you integrate review commits into other ones, through git
rebase -i fixup/squash operations, and only then, the appropiate
Helped-by trailers will be automatically inserted.

You need a git version that supports interpret-trailers for this to
work. You should also set `git config trailer.ifexists addIfDifferent`.

elmart committed:
Review trailers tooling: README.

bfredl commented on contrib/review-trailers-tooling/NVIM_RUNTIME_PATH/plugin/insert-github-user-info.vim:
```diff
@@ -0,0 +1,14 @@
+" Replace word under cursor (which should be a GitHub username)
+" with some user info ("Full Name <email@address>").
+" If info cout not be found, "Not found" is inserted.
+function! <SID>InsertGitHubUserInfo()
```
Nice! why not export the function, so it easily can be remapped? Also slight issue, it doesn't work if username is prefixed with `@`

elmart replied:
Yes, we can polish this a bit. This was just a quick-and-dirty way of having useable.

Pyrohh replied:
I barely know vimscript, but does this accept a range? I'd think it would be quite useful if many `Helped-by:` fields were present.

elmart replied:
Well, certainly, range handling could be added. But note that, in principle, this is not intended to edit existing trailers, which shouldn't be there if they only have the username. Rather, this is intended to enter new trailers, where you type the trailer title (or use provided UltiSnips templates to do that), and then, enter a username, and quickly transform it into fullname and email. So, I'm not sure if range handling has real sense in such a context.

bfredl commented on contrib/review-trailers-tooling/README.md:
```diff
@@ -0,0 +1,41 @@
+- `Helped-by: Full Name <email@address>`
+   Added by regular users to particular commits within a PR. Applies to the commit it appears on. Means "this commit was somehow improved based on suggestions by @reviewer".
+- `Reviewed-by: Full Name <email@address>`
+   Added by maintainer to PR merge commit. Applies to whole PR. Means "these changes were carefully reviewed by @reviewer an he has stated conformity with them."
```
"an he has": spelling, and I think we should use gender-neutral pronoun.

elmart replied:
Ok. Regarding gender-neutrality, well, I'm ok with it too. When I learned English, I was taught that "generic he" was the correct thing. Nowadays, though, it seems "singular they" is having more adepts.I don't have a strong position on that. I just think it's a bit of a pity some people make an argument out of that. To me, offense is in the intention, not in the words you use. But, again, I don't have any problem in using singular they, if that's considered to be better here.

bfredl replied:
> I just think it's a bit of a pity some people make an argument out of that. To me, offense is in the intention, not in the words you use.

In general I agree with this. I was just assuming, if we agree trailers are the way to go, this paragraph would be moved to `CONTRIBUTING.md` or similar (as is not really related to this tool), at which point it would be a "policy" document and not "personal communication" anymore, and a bit more care might be warranted. But not "big issue" or anything, just my opinion :)

elmart replied:
No, no, you're completely right. I'm talking about workflow here because what tools should do can depend on what the intended workflow is. But, when we reach an agreement on that, workflow description should be moved to wiki, as you mention.

Pyrohh commented on contrib/review-trailers-tooling/GIT_DIR/hooks/post-rewrite:
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
```
Probably should be `/usr/bin/env bash` (for FreeBSD, among others).

Pyrohh replied:
Unless git does some magic here...

elmart replied:
No, I don't think so. You're right using `env` is more portable.

Pyrohh commented on contrib/review-trailers-tooling/PATH/github_user_info:
```diff
@@ -0,0 +1,61 @@
+
+get_from_github()
+{
+    IFS='|' read fullname email < <(curl -s "$users_api/$username" |
```
`curl -Ss`? Or do we not care if curl fails here?

elmart replied:
We don't care. If `curl ` fails, `fullname` will be empty, and the rest will work the same.

Pyrohh commented on contrib/review-trailers-tooling/PATH/github_user_info:
```diff
@@ -0,0 +1,61 @@
+
+get_from_log()
+{
+    result=$(git log --author="$fullname" -1 |
```
In case the user uses `color.ui = always`, then maybe `--color=never` too? I'm not totally sure it's needed.

elmart replied:
I don't think color is used when outputting to a non-terminal. I'll check anyway.

Pyrohh commented on contrib/review-trailers-tooling/README.md:
```diff
@@ -0,0 +1,41 @@
+## What's this?
+
+A basic set of tools to ease adding review trailers, in order to acknowledge reviewers' work.
```
Maybe add a link to something like https://git.wiki.kernel.org/index.php/CommitMessageConventions? I don't think trailers are too common, but that might just be my experience.

elmart replied:
Ok. Maybe http://git-scm.com/docs/git-interpret-trailers better?

Pyrohh replied:
Well Git's man pages are pretty dense, but the one you mentioned doesn't seem that bad.

On March 29, 2015 5:35:04 PM EDT, "Eliseo Martínez" <notifications@github.com> wrote:
>> @@ -0,0 +1,41 @@
>> +## What's this?
>> +
>> +A basic set of tools to ease adding review trailers, in order to
>acknowledge reviewers' work.
>
>Ok. Maybe http://git-scm.com/docs/git-interpret-trailers better?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2291/files#r27359403

Pyrohh commented on contrib/review-trailers-tooling/README.md:
```diff
@@ -0,0 +1,41 @@
+
+- Command line tool to transform GitHub usernames into `Full Name <email@address>` info.
+  Script uses GitHub API to look for the info. If user doesn't have a published email, it tries to find a matching one at local git repo. Successful results are cached locally to speed up subsequent requests for the same user.
+  INSTALL: You need curl and jq for this work. Then, put the script in your path.
```
perhaps surround `curl` and `jq` with backtics, and provide a link to their home pages:

https://stedolan.github.io/jq/
http://curl.haxx.se/

elmart replied:
:+1:

Pyrohh commented on contrib/review-trailers-tooling/README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Which tools are included, and how they help?
+
+- Command line tool to transform GitHub usernames into `Full Name <email@address>` info.
```
maybe Command line tool -> Command line tools?

elmart replied:
That refers to `github-user-info` executable, which is a single command-line tool.

Pyrohh replied:
Ah, I see. In that case I think `A command-line tool ...` works better.

On March 29, 2015 5:36:37 PM EDT, "Eliseo Martínez" <notifications@github.com> wrote:
>> @@ -0,0 +1,41 @@
>> +## What's this?
>> +
>> +A basic set of tools to ease adding review trailers, in order to
>acknowledge reviewers' work.
>> +
>> +## Which trailers are to be added?
>> +
>> +- `Helped-by: Full Name <email@address>`
>> +   Added by regular users to particular commits within a PR. Applies
>to the commit it appears on. Means "this commit was somehow improved
>based on suggestions by @reviewer".
>> +- `Reviewed-by: Full Name <email@address>`
>> +   Added by maintainer to PR merge commit. Applies to whole PR.
>Means "these changes were carefully reviewed by @reviewer an he has
>stated conformity with them."
>> +
>> +## Which tools are included, and how they help?
>> +
>> +- Command line tool to transform GitHub usernames into `Full Name
><email@address>` info.
>
>That refers to `github-user-info` executable, which is a single
>command-line tool.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2291/files#r27359417

Pyrohh commented on contrib/review-trailers-tooling/NVIM_RUNTIME_PATH/UltiSnips/gitcommit.snippets:
```diff
@@ -0,0 +1,14 @@
+Resolution :
+endsnippet
+
+snippet "^r" "Insert Reviewed-by trailer" rb
```
Maybe just `"Reviewed-by trailer"`? I'd think that `Insert` in a snippet description is redundant (same issue with the snippet below, if it is an actual issue).

elmart replied:
:+1:

justinmk commented:
>  tool to transform GitHub usernames into corresponding "Full Name <email@addres>".

That sounds useful. 

I don't use snippets though.

elmart commented:
@justinmk 

The think I'm most interested about regarding your feedback is proposed workflow (what do trailers mean, who adds them, and when.

What I'm currently proposing is:
- `Helped-by: Full Name <email@address>`
   Added by regular users to particular commits within a PR. Applies to the commit it appears on. Means "this commit was somehow improved based on suggestions by @reviewer".
- `Reviewed-by: Full Name <email@address>`
   Added by maintainer to PR merge commit. Applies to whole PR. Means "these changes were carefully reviewed by @reviewer an he has stated conformity with them."

This has the drawback of imposing some extra work on maintainers/mergers, as they have to add the `Reviewed-by` trailers. On the other side, it has the benefit of completely automating addition of `Helped-by` trailers by regular users.

But other approaches are possible. For example: Make `Reviewed-by` trailers be added by regular users, too. This has the clear benefit of removing that work from maintainer's shoulders. But, then, `Reviewed-by` trailer can't be done on the merge commit, as regular users don't have access to it. So, if we do it this way, user should have to add the `Reviewed-by` trailer to all commits within a PR.

I could easily modify provided tools to support such other workflow. For example, I could modify provided git hook so that, if you invoke rebase with REVIEWERS env var set to a list of usernames, the corresponding `Reviewed-by` trailers get automatically added to every commit, in addition to maintaining the previous behaviour of adding `Helped-by` trailers to particular commits that received a fixup.

Or I could git aliases to support commands like `git reviewed-by <reviewer>... <start_point>` and `git helped-by <reviewer>...`. 

Possibilities are many. I just posted the current changes to have something to start discussion with. But I need for us to reach an agreement on trailer meaning and who/when adds them, before polishing the tools to smoothy support the chosen workflow

@Pyrohh @bfredl 

I will address all your suggestion as soon as we have an agreement on what said above.

Shougo commented:
I think it is better to contain neosnippet/snipMate snippet file.
And it should be distributed one plugin package.

justinmk commented:
Can revisit if this comes up in the future.
